### PR TITLE
run: Fix up various things.

### PR DESCRIPTION
**Pull request by PhilipMetzger**
Adress the post-merge comments from #2486, which found a doc comment inaccurate and to not blindly ignore `-j 0` which would've worked until now. I've also reduced the default `jobs` size to one, as it's user-visible configuration
which determines how many processes should run.

Thanks to @necauqua the controversial `unsafe` usage was already removed.

I've omitted to change `revisions` from an Vec to a RevisonArg for the moment,  as I will keep working on the file anyway.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/commands/run.rs [hunk 1]
```diff
@@ -55,16 +53,16 @@ pub fn cmd_run(ui: &mut Ui, command: &CommandHelper, args: &RunArgs) -> Result<(
...
+        .jobs
+        .filter(|j| *j > 0)
+        .or_else(|| {
+            // Use all available cores
+            std::thread::available_parallelism()
+                .ok()
+                .and_then(|t| t.try_into().ok())
+        })
+        // Fallback to two cores as we don't know on which system we're running on.
+        .unwrap_or(2usize);
```
Comment (martinvonz) [inline cli/src/commands/run.rs:66]: I still prefer to default to 1 job, regardless of number of cores.

Comment (arxanas) [reply cli/src/commands/run.rs:66]: - I'm a fan of 1 job by default.
- How are you actually going to use two threads for `jj run`? Do you need to spawn a new workspace to run the extra job? (Maybe we should have left the `--jobs` argument out until we could actually use it.)

Comment (PhilipMetzger) [reply cli/src/commands/run.rs:66]: Thanks for making me look up the implementations for both `tokio::process` and `tokio::fs`. In both platform implementations they suggest that they can't really a process asynchronously (e.g `epoll` / `iocp`), so they listen on best effort for signals. And for creating directories, which is needed for a certain scale of commits, they will be spawned on a background job anyway. 

Initially I partially agreed with one job, but after looking into it, I'd rather do it myself, so we have atleast one controlling thread free.

Comment (martinvonz) [reply cli/src/commands/run.rs:66]: Does the flag control how many concurrent subprocesses we allow or the number of threads to use? I think we should give the user control over the former, but hopefully we don't need to give them control over the latter.

Comment (PhilipMetzger) [reply cli/src/commands/run.rs:66]: > Does the flag control how many concurrent subprocesses we allow or the number of threads to use? I think we should give the user control over the former, but hopefully we don't need to give them control over the latter.

I'd say the latter, as a subprocess w/o a existing directory to work in is useless. Why do you think that controlling processes will help the user? I imagine a user with a threadripper or a beefy 128 core machine will want it done in short a time.

Comment (martinvonz) [reply cli/src/commands/run.rs:66]: See https://github.com/martinvonz/jj/pull/2486#discussion_r1382266739. I didn't understand your reply there.

Comment (PhilipMetzger) [reply cli/src/commands/run.rs:66]: External handholding was always out of scope until now.

Comment (PhilipMetzger) [reply cli/src/commands/run.rs:66]: > * How are you actually going to use two threads for `jj run`? Do you need to spawn a new workspace to run the extra job? 

I'd like to have a thread to control the creation of commit trees.

> (Maybe we should have left the `--jobs` argument out until we could actually use it.)

I'd rather not wait longer on a implementation, as this is incremental in-tree development. We can still change _it_ before users can actually use the command. 

The rest of my replies are inline as edits.

Comment (arxanas) [reply cli/src/commands/run.rs:66]: I'm fine to proceed with `--jobs` as-is, since we don't use it for anything yet, anyways, but I think there's a fundamental mismatch in meaning about "parallelism":

- One notion is the maximum number of concurrent instances of the user-provided command to run. This corresponds to `git-branchless test`'s `--jobs` argument.
- One notion is the size of some thread pool used to service those jobs, as well as other maintenance overhead, such as checking out a workspace on disk. I don't think the user typically *wants* to configure that value themselves, and, when so, they would probably want to configure it using a more general mechanism that lets you set the size of jj thread pools for various commands.

The design document does not explicitly specify what "parallelism" means, but I would have expected it to mean the former, which means that a thread in control of creating commit trees would not count towards/be relevant to the `--jobs` flag.

Comment (PhilipMetzger) [reply cli/src/commands/run.rs]: Ok I've adjusted it to a single user visible job. The implementation details may use a single or multiple threads, but that's open anyway.

File: cli/src/commands/run.rs [hunk 2]
```diff
@@ -55,16 +51,16 @@ pub fn cmd_run(ui: &mut Ui, command: &CommandHelper, args: &RunArgs) -> Result<(
...
+        .jobs
+        .filter(|j| *j > 0)
+        .or_else(|| {
+            // Use all available cores
+            std::thread::available_parallelism()
+                .ok()
+                .and_then(|t| t.try_into().ok())
+        })
+        // Fallback to a single user-visible job.
+        .unwrap_or(1usize);
```
Comment (arxanas) [inline cli/src/commands/run.rs:64]:
```suggestion
    let jobs = match args.jobs {
        Some(0) | None => std::thread::available_parallelism().ok().and_then(|| t.try_into().ok()),
        Some(jobs) => Some(jobs),
    };
    let _jobs = jobs.unwrap_or(1usize);
```

- Personally, I think the use of `filter` here is hard to follow.
- After rewriting the code as in my suggestion, I realize that I think it doesn't do what we want: if the user doesn't pass `--jobs`, we should default to `1`, not the max parallelism. That is, we should only match on `Some(0)` for the max-parallelism case. (We don't want to e.g. spawn extra workspaces without the user explicitly opting-in, since that could use a lot of resources.)

**Review by arxanas — APPROVED** [review]
Accepting to unblock.

File: cli/src/commands/run.rs (hunk 2 above)
Comment (PhilipMetzger) [reply cli/src/commands/run.rs:64]: > * Personally, I think the use of `filter` here is hard to follow.

I'll add a comment to hopefully increase readability. 

> * After rewriting the code as in my suggestion, I realize that I think it doesn't do what we want: if the user doesn't pass `--jobs`, we should default to `1`, not the max parallelism. That is, we should only match on `Some(0)` for the max-parallelism case. (We don't want to e.g. spawn extra workspaces without the user explicitly opting-in, since that could use a lot of resources.)

I still disagree with only spawning a single "job", if we pass nothing for `-j`. I'm going to drop the argument now, but I'd rather spawn a "job" for each core then just default to one and if all else fails we use a single job. Your argument implies dropping `std::thread::available_parallelism()` completly. (I put job in quotes, as it probably is a process in user-terms and still implies something thread-like to me.)

The one thing I realised is that I want actual parallelism for the MVP of`run`, not concurrency. At some point in the future we then can rewrite it in terms of concurrency.

Comment (PhilipMetzger) [reply cli/src/commands/run.rs]: I've added a comment for the `filter` part.

**Comment by PhilipMetzger** [comment]
I'd like to finally land this patch tomorrow, so call for last reviews or nits.

File: cli/src/commands/run.rs [hunk 3]
```diff
@@ -55,16 +51,21 @@ pub fn cmd_run(ui: &mut Ui, command: &CommandHelper, args: &RunArgs) -> Result<(
...
-        available.into()
-    };
+    // Jobs are resolved in this order:
+    // 1. Commandline argument iff > 0.
+    // 2. the amount of cores available.
+    // 3. a single job, if all above fails.
+    let _jobs = args
+        .jobs
+        // We need to fixup a `jj run -j0` input, as 0 jobs is a invalid state.
+        .filter(|j| *j > 0)
```
Comment (martinvonz) [inline cli/src/commands/run.rs:61]: We generally treat an invalid argument as an error. Should we do that here too?

**Review by martinvonz — APPROVED** [review]
